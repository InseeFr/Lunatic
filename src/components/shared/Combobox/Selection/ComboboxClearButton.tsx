import { useMemo } from 'react';
import classnames from 'classnames';
import { CrossIcon } from '../../Icons';
import { slottableComponent } from '../../HOC/slottableComponent';
import { IconButton } from '../../Button/IconButton';

function isDisabled(search?: string) {
	return !search || search.trim().length === 0;
}

function createOnKeyDown(onClick?: () => void) {
	if (typeof onClick !== 'function') {
		return undefined;
	}
	return (e: { key: string }) => {
		if (e.key === 'Enter') {
			onClick();
		}
	};
}

type Props = {
	className?: string;
	search?: string;
	onClick?: () => void;
	editable?: boolean;
};

function LunaticComboboxClearButton({
	className,
	search,
	onClick,
	editable,
}: Props) {
	const onKeyDown = useMemo(() => createOnKeyDown(onClick), [onClick]);
	if (!editable) {
		return null;
	}
	return (
		<IconButton
			className={classnames('mini', 'lunatic-combo-box-fab', className)}
			title="delete"
			onClick={onClick}
			disabled={isDisabled(search)}
			onKeyDown={onKeyDown}
		>
			<CrossIcon className="lunatic-combo-box-icon" />
		</IconButton>
	);
}

export const ComboboxClearButton = slottableComponent(
	'ComboboxClearButton',
	LunaticComboboxClearButton
);
