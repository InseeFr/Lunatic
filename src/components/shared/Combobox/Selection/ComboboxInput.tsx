import React, {
	useRef,
	useEffect,
	type HTMLAttributes,
	type KeyboardEventHandler,
} from 'react';
import classnames from 'classnames';
import { slottableComponent } from '../../HOC/slottableComponent';

export type InputProps = {
	placeholder?: string;
	disabled?: boolean;
	value?: string;
	labelledBy?: string;
	focused?: boolean;
	invalid?: boolean;
	readOnly?: boolean;
} & HTMLAttributes<HTMLInputElement>;

function LunaticComboboxInput({
	placeholder,
	disabled,
	onChange,
	value,
	id,
	labelledBy,
	focused,
	className,
	invalid,
}: InputProps) {
	const inputEl = useRef<HTMLInputElement>(null);

	useEffect(() => {
		if (inputEl.current && focused) {
			inputEl.current.focus();
		}
	}, [inputEl, focused]);

	const onKeydown: KeyboardEventHandler = (e) => {
		const { key } = e;
		if (key === 'ArrowUp' || key === 'ArrowDown') {
			e.preventDefault();
		}
	};

	return (
		<input
			ref={inputEl}
			id={id}
			className={classnames('lunatic-combo-box-input', className)}
			type="text"
			onChange={onChange}
			value={value}
			aria-invalid={invalid}
			title="combo-box"
			autoComplete="off"
			autoCapitalize="off"
			autoCorrect="off"
			spellCheck="false"
			placeholder={placeholder}
			disabled={disabled}
			aria-labelledby={labelledBy}
			onKeyDown={onKeydown}
		/>
	);
}

export const ComboboxInput = slottableComponent(
	'ComboboxInput',
	LunaticComboboxInput
);
