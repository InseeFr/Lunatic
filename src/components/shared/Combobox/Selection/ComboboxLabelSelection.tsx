import classNames from 'classnames';
import type { ComboboxOptionType } from '../ComboboxType';
import { slottableComponent } from '../../HOC/slottableComponent';

type Props = {
	option?: ComboboxOptionType | null;
	placeholder?: string;
	search?: string;
	disabled?: boolean;
	readOnly?: boolean;
};

function getContent(
	option: Props['option'],
	search?: Props['search'],
	placeholder?: string
) {
	if (option) {
		const { id, value, label } = option;
		return label ? `${id || value} - ${label}` : id || value;
	}
	if (search && search.trim().length) {
		return search;
	}
	return placeholder ?? '';
}

const LunaticComboBoxLabelSelection = ({
	option,
	placeholder,
	search,
	disabled = false,
	readOnly = false,
}: Props) => {
	const isPlaceholder = !option && (!search || search.length === 0);
	if (option?.value === 'OTHER') {
		return (
			<div
				className={classNames('lunatic-combo-box-selected', {
					disabled,
					readOnly,
				})}
			>
				<span className="selection">{option.label}</span>
			</div>
		);
	}
	return (
		<div
			className={classNames('lunatic-combo-box-selected', {
				disabled,
				readOnly,
			})}
		>
			<span
				className={classNames({
					placeholder: isPlaceholder,
					selection: !isPlaceholder,
				})}
			>
				{getContent(option, search, placeholder)}
			</span>
		</div>
	);
};

export const ComboboxLabelSelection = slottableComponent(
	'ComboboxLabelSelection',
	LunaticComboBoxLabelSelection
);
