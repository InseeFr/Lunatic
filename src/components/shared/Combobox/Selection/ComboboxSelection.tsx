import classnames from 'classnames';
import { ComboboxLabelSelection } from './ComboboxLabelSelection';
import { ComboboxInput } from './ComboboxInput';
import type { ComboboxSelectionProps } from '../ComboboxType';
import type { ChangeEvent } from 'react';

/**
 * Label displayed when a value is selected in the ComboBox
 * This label can either be an input (when editable or expanded) or a simple span
 */
export function ComboboxSelection({
	labelRenderer,
	placeholder,
	search,
	expanded,
	disabled,
	readOnly,
	focused,
	onChange,
	selectedIndex,
	options,
	editable,
	labelId,
	id,
	classNamePrefix,
	invalid,
}: ComboboxSelectionProps) {
	const showLabel = !editable || !expanded;
	const selectedOption =
		selectedIndex !== undefined ? options[selectedIndex] : undefined;
	const LabelSelectionComponent = labelRenderer ?? ComboboxLabelSelection;

	return (
		<div
			id={id}
			className={classnames(
				`${classNamePrefix ?? 'lunatic'}-combo-box-selection`,
				{
					focused,
					disabled,
				}
			)}
			role="combobox"
			aria-controls={'todo'}
			aria-haspopup="listbox"
			aria-expanded={expanded}
			aria-autocomplete="list"
			aria-owns={id}
			aria-labelledby={labelId}
		>
			{showLabel ? (
				<LabelSelectionComponent
					option={selectedOption}
					placeholder={placeholder}
					search={search}
					disabled={disabled}
					readOnly={readOnly}
				/>
			) : (
				<ComboboxInput
					invalid={invalid}
					id={`combobox-input-${id}`}
					className="lunatic-combo-box-input"
					onChange={(e: ChangeEvent<HTMLInputElement>) =>
						onChange?.(e.target.value)
					}
					value={search}
					placeholder={placeholder}
					disabled={disabled}
					readOnly={readOnly}
					focused={focused}
					labelledBy={labelId}
				/>
			)}
		</div>
	);
}
