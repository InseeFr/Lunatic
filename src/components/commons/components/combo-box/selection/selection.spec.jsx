import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Selection from './selection';
import { expect, describe, it, vi } from 'vitest';
const options = [
	{ label: 'Apple', value: 'apple' },
	{ label: 'Orange', value: 'orange' },
	{ label: 'Banana', value: 'banana' },
];

describe('Selection component', () => {
	it('should render correctly with default props', () => {
		const { container } = render(
			<Selection
				labelRenderer={({ option }) => (
					<div>{option ? option.label : 'No selection'}</div>
				)}
				options={options}
			/>
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render correctly when it is editable', () => {
		const { container } = render(
			<Selection
				labelRenderer={({ option }) => (
					<div>{option ? option.label : 'No selection'}</div>
				)}
				options={options}
				editable
			/>
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render correctly when it is expended', () => {
		const { container } = render(
			<Selection
				labelRenderer={({ option }) => (
					<div>{option ? option.label : 'No selection'}</div>
				)}
				options={options}
				expended={true}
			/>
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render correctly when it is disabled', () => {
		const { container } = render(
			<Selection
				labelRenderer={({ option }) => (
					<div>{option ? option.label : 'No selection'}</div>
				)}
				options={options}
				disabled={true}
			/>
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	it('should render the selected option when there is a selection', () => {
		const { getByText } = render(
			<Selection
				labelRenderer={({ option }) => (
					<div>{option ? option.label : 'No selection'}</div>
				)}
				options={options}
				selectedIndex={0}
			/>
		);

		expect(getByText('Apple')).toBeInTheDocument();
	});

	it('should not render the input when it is not editable', () => {
		const { queryByPlaceholderText } = render(
			<Selection
				labelRenderer={({ option }) => (
					<div>{option ? option.label : 'No selection'}</div>
				)}
				options={options}
				placeholder="Search fruits"
				editable={false}
			/>
		);

		expect(queryByPlaceholderText('Search fruits')).not.toBeInTheDocument();
	});
});
