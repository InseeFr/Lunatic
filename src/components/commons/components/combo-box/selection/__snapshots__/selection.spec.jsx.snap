// Vitest Snapshot v1

exports[`Selection component > should render correctly when it is disabled 1`] = `
<div
  aria-autocomplete="list"
  aria-controls="todo"
  aria-haspopup="listbox"
  aria-owns="idUndefined"
  class="lunatic-combo-box-selection"
  id="idUndefined"
  role="combobox"
>
  <div
    class="lunatic-combo-box-selected disabled"
  >
    <div>
      No selection
    </div>
  </div>
</div>
`;

exports[`Selection component > should render correctly when it is editable 1`] = `
<div
  aria-autocomplete="list"
  aria-controls="todo"
  aria-haspopup="listbox"
  aria-owns="idUndefined"
  class="lunatic-combo-box-selection"
  id="idUndefined"
  role="combobox"
>
  <div
    class="lunatic-combo-box-selected"
  >
    <div>
      No selection
    </div>
  </div>
</div>
`;

exports[`Selection component > should render correctly when it is expanded 1`] = `
<div
  aria-autocomplete="list"
  aria-controls="todo"
  aria-haspopup="listbox"
  aria-owns="idUndefined"
  class="lunatic-combo-box-selection"
  id="idUndefined"
  role="combobox"
>
  <div
    class="lunatic-combo-box-selected"
  >
    <div>
      No selection
    </div>
  </div>
</div>
`;

exports[`Selection component > should render correctly with default props 1`] = `
<div
  aria-autocomplete="list"
  aria-controls="todo"
  aria-haspopup="listbox"
  aria-owns="idUndefined"
  class="lunatic-combo-box-selection"
  id="idUndefined"
  role="combobox"
>
  <div
    class="lunatic-combo-box-selected"
  >
    <div>
      No selection
    </div>
  </div>
</div>
`;
